### ISSUE TITLE
Align on ideal where & how for Account Verification

### ISSUE THREAD

**Issue by m-branson** [mozilla-iam/mozilla-iam#58]
I am working to better determine where & how account verification should *ideally* take place both technically & for our users.

As a result of our latest implementation of Mozillians.org `Profile Identitites` and this subsequent [NLX issue](https://github.com/mozilla-iam/auth0-custom-lock/issues/168), it has become clear that there is misalignment re: the ideal technical solution. In order to iterate on the user experience, I'd like to better understand where we are headed technically; e.g. `the code` 😏 

This **NOTE** from @gdestuynder (via the above mentioned issue) is the first I've realized that Mozillians.org *is the preferred account verification point* and that the current method of routing through NLX is a *technical work around*.

>NOTE: our method of doing account verification still utilizes our authentication stack and creates users automatically in Auth0. The method we prefer, and that apparently will be used for Bugzilla verification, is to directly verify through Mozillians.org instead of requesting a login (except for LDAP, which is correctly/technically a login request)
>
>I just want to make it clear that we ideally want to stop doing that, so the work above are still work around (though, it looks great otherwise and make its a much nicer work-around UX wise).
>
>I'm adding this note in case this changes your priorities or anything like that.

### The Questions:

1. What is our preferred technical solution?
2. Can someone please map out this technical solution?
3. Is the entire team aligned on this technical solution?
4. How does `login.mozilla.com/`factor into this verification model? 
   - Will LDAP Password & 2FA still be managed here?
   - Is there potential to make this `Account Management Portal` the location for all `Profile Identity Account Verification`? Or will LDAP specific auth be migrated into Mozillians.org as well?
   - From a UX standpoint I'm motivated to provide users a focused location to handle all of their account verification needs. If that is Mozillians.org, great. If not, where is it?

Happy to take this to tech alignment as well, would welcome some async clarification here in the meantime.

cc @hmitsch @hidde @gene1wood @akatsoulas @andrewkrug @johngian @jeffbryner

**Comment by hmitsch** [comment]
Some answers here are simple, others not so much ... let's start with simple ... I will try to answer in separate comments to allow for easier quasi-threading and referencing.

**Comment by hmitsch** [comment]
In the long run, `login.mozilla.com` should go away. The _repositioned Mozillians_ is a great place to hold most (all) of its functionality.

**Comment by gdestuynder** [comment]
1. Mozilians.org custom code calls various IdPs we want to verify accounts with. These are not necessarily accounts you can use to login to Mozilla IAM, it could be any account (Microsoft Live, GitHub, GitLab, etc.) that we want to support. Most of the time it's an OAuth2 call.

2. This is part of @akatsoulas work on Bugzilla's integration for this so I'll skip it. It's not very complex from a technical pov.

3. Yes, we met on this at the All hands

4. login.mozilla.com currently does not matter for account verification
  - LDAP password and 2FA are still manager there, not related
  - Authentication to Mozilla IAM is  **separated** from account verification. This is very important. They are different concepts in this case, even thus right now for a user they all look like you're login into something (which part of the UX issue).
 - Mozillians.org is where account **verification** happens. Not 2FA or password management.

Hope this helps

**Comment by hmitsch** [comment]
> 1. What is our preferred technical solution?
> 3. Is the entire team aligned on this technical solution?

My understanding is that the solution @gdestuynder describes is:
* **Yes**, it's our preferred technical solution.
* **Yes**, the entire IAM Team is aligned on this solution.

**Comment by gdestuynder** [comment]
> In the long run, login.mozilla.com should go away. The repositioned Mozillians is a great place to hold most (all) of its functionality.

Let's have this discussion at a later time as it can be confusing. We do not own most IdP and the password/2FA management of the IdP will *never* be in mozillians.org as it's simply impossible.

The one exception is LDAP because we own it, but there is currently **no** plan to change this. We'll probably do this as a microservice that looks like its part of Mozillians.org from a UX pov but not from a technical pov. I.e. LDAP password management and 2FA management will not be in the Mozillians.org django code base. It also does a bunch of other things such as vpn, wifi certificates.

In other words, again, I think it's better to just ignore this part for now  so that we're not confused, and explore it in 2019.

**Comment by flekkowich** [comment]
I have a question around what it means from a RP perspective that mozillans.org becomes preferred. I saw great user value in having that 'less company branded' but let's discuss.... I'm curious is this was included in the discussions that got us there. But we can take that at next IAM tech alignment meeting. Loads to think about!

**Comment by hmitsch** [comment]
Attached is a V2 of the above sketch. It incorporates @m-branson's feedback:
* Replaced ~add identity~ by **verify** 
* Google and LDAP added
* (slight) reference to LDAP-via-NLX verification

![2018-06-29 13 17 00](https://user-images.githubusercontent.com/869692/42089784-0c984106-7b9f-11e8-8320-741c97b0d769.png)

Hope this makes more sense. Happy about feedback.

**Comment by fiji-flo** [comment]
I'd like us to take a step back:

From my point of view we need to distinguish between **identity providers** and **verified accounts**. Right now we treat them the same, but not. In my language an **identity provider** should be something a user can **log in** with. Whereas, a **verified account** is an account that we know the user has/had control over.

Currently we only have the concept of **identity providers**, but with a hidden ranking. And if you add a lower ranked identity provider it's basically just a **verified account** (a user cannot use it to login / can login but has not access).

- Should we reflect the concept of these two different things **identity providers** and **verified accounts** in our new UX?
- Should we only allow a single **identity provider** for a user (and how to change that / fall back)?
  - Should this then be reflected in auth0?
  - Mozillians could have an account recovery based on verified accounts.
- We should force them to overlap, the **identity provider** will automatically be a **verified account**.

In terms of login.mozilla.com: I agree with @gdestuynder and we should rather have a separate page with aligned UX and clearly link to it.

**Comment by hmitsch** [comment]
> * Should we reflect the concept of these two different things identity providers and verified accounts in our new UX?

**Yes!**

> * Should we only allow a single identity provider for a user? 

Yes, this is today's implementation. I'd like to avoid changing _all the things_ as we work towards Phase 1 of _Mozillians Repositioning_.

> (and how to change that / fall back)

This is today's _customer support headache_. I suggest we keep today's implementation and find smart ways to fix this on the experience level.

> * We should force them to overlap, the identity provider will automatically be a verified account.

Yes, totally agree with that.

Lastly, I want to re-state that I also agree with @gdestuynder: login.m.c is a topic we certainly want to defer.

**Comment by flekkowich** [comment]
**This point is separate from overall discussion** 
Jumping back in again @m-branson .... 

what I mean is the following - (this is my mental model right now, but it may be wrong!)

Current state (as per Henrik's drawing) we have NLX as identity verification with RPs - this means a RP like Common Voice can create it's own profiles and use NLX to handle the account verification. A huge win for RPs b/c they don't have to worry about security (and I assume other things as well)? And something other RPs will most likely see huge wins in. 

My mental model was that we would 'somehow' create a NLX instance of identity managment (yes this would be 'branded' Mozilla, but similar to NLX UI overall, not in a overly Mozilla kinda way. If we move identity management to Mozillans.org (pending name) this makes the connection to Mozilla / Mozilla profile much clearer and so could 'turn' some RPs away. 

Many assumptions ins this statement, but this is my concern... @hmitsch is this something you have considered? Have I misunderstood something completely (very likely!)

**Comment by fiji-flo** [comment]
@flekkowich 
> Current state (as per Henrik's drawing) we have NLX as identity verification with RPs - this means a RP like Common Voice can create it's own profiles and use NLX to handle the account verification.
Again, verification and login are two different things. In the common voice we use auth0 as an SSO provider via 3rd party authentication providers. This is for authentication.

If we speak of identity management there are two things:
- Identity in form of what identity providers can I use for SSO
- Which verified online identities do I have
Both of them can overlap heavily which makes this so complicated to communicate.

From what I understand you want us to manage the identities someone can use to as SSO login?

I agree that we should defer this, but I also really the issue here. What we can do is document the current state and make sure we and our users can understand this. That would also help to think of a brighter future for this topic.

**Comment by m-branson** [comment]
Fantastic, this is the level of alignment I was hoping to understand re: IdP vs. Verified Accounts...

>        Should we reflect the concept of these two different things identity providers and verified accounts in our new UX?
>
>Yes!

100% agree, very much want these to be communicated as separate things.

>        Should we only allow a single identity provider for a user?
>
>Yes, this is today's implementation. I'd like to avoid changing all the things as we work towards Phase 1 of Mozillians Repositioning.

Noted.

>    (and how to change that / fall back)
>
>This is today's customer support headache. I suggest we keep today's implementation and find smart ways to fix this on the experience level.

Agreed, this is a headache and a major goal of this thread: `"From a UX standpoint I'm motivated to provide users a focused location to handle all of their account verification needs."` This includes that ability to edit / fall back.

Knowing that we have a 'line in the sand' re: Mozillians.org being the place for this I can map accordingly & start working toward the _ideal_ experience.

>        We should force them to overlap, the identity provider will automatically be a verified account.
>
>Yes, totally agree with that.

100% agreement. If an IdP is connected, it should also reflect in the front-end as added to the verified accounts section as well.

**Comment by gdestuynder** [comment]
Another note, we've created https://github.com/mozilla-iam/mozilla-iam/blob/master/GLOSSARY.md#account-verification which defines the terms / serves as reference (and can be updated, just click the pen icon and PR a change :)

In this case "account verification" is defined, and i like to think of it as "proof of ownership"

**Comment by m-branson** [comment]
@fiji-flo re: this...

>From what I understand you want us to manage the identities someone can use to as SSO login?

This is how I understand what Rina was saying as well.

@flekkowich re: this...

>My mental model was that we would 'somehow' create a NLX instance of identity managment (yes this would be 'branded' Mozilla, but similar to NLX UI overall, not in a overly Mozilla kinda way. If we move identity management to Mozillans.org (pending name) this makes the connection to Mozilla / Mozilla profile much clearer and so could 'turn' some RPs away.

This Identity Provider (IdP) management option -- a separate portal where Account Verification occurs  vs. solely routing users to Mozillians.org to manage their IdP -- is what I was looking to understand as well.

If I use Common Voice as an example and compare @flekkowich's mental model vs. @hmitsch's mental model, this is how it's mapped:

### IdP Portal method
![idp-method](https://user-images.githubusercontent.com/18686986/42108601-ddcc038a-7b8f-11e8-8830-2e968e564263.JPG)


### Mozillians.org IdP method
![mozillians-method](https://user-images.githubusercontent.com/18686986/42108558-c1305014-7b8f-11e8-92d9-378dd7283777.JPG)

On the note of branding & user impact; the IdP Portal method is the path of least resistance IMO. 
1. It would allow RPs to link to IdP management without forcing use of a Mozillians.org profile
2. It would allow users who prefer to manage their IdP via their Mozillians.org profile to do so as well
3. Ultimately this _could_ allow for more flexibility overall and still provide a central profile for those Mozillians who prefer to have one.

Thoughts? Edits?

**Comment by fiji-flo** [comment]
> This Identity Provider (IdP) management option -- a separate portal where Account Verification…

Again this would just manage SSO login providers and not all verified accounts. One example will be bugzilla. User can verify that they are the owner of a bugzilla account but not login with it. This account should not be in this IdP.

However this makes everything a little bit messy. I my personal ideal world the IdP is part of Mozillians / our profile management tool. This would mean every common voice user also has a profile (which is not visible to anyone unless the user *upgrades* it to a Mozillians profile).

However this is for away from what we designed. This would need a core profile (identity management + little user data) and everything is built on top of that. I think this is basically what auth0 offers, but we don't want to buy into that fully.

**Comment by gdestuynder** [comment]
Could the "IdP portal" simply be Mozillians.org with the UX refresh, so that it looks like its integrated?

We only allow *one* entity to edit a profile field (ie each field is owned and controlled by one entity), which means if common voice had its own way of setting the user's account verifications, then mozillians.org wouldn't be able to do it anymore.

This is because otherwise we run into issues where different parties disagree on the field/attribute values.

**Comment by andrewkrug** [comment]
For me the IdP portal should be in Mozillians since this is a core part of what IAM is doing.  Other identities you have verified are all attributes of you as a mozillian.  I would caution against us adding too much scope here.  

### For example:

* Today we can only verify oauth providers.  This means without continuing to oversubscribe the auth0 infrastructure we can not verify LDAP accounts with an oauth call.
* Today we only have need to verify Github, Bugzilla, and FxA.  There is no current proposed use case for a secure assertion including a Mozillian's twitter account for example.  

When I see the Mozillians.org IdP method diagram and it contains so many login methods that adds a lot of code that will have to be maintained over time.  

A lot of this is also caught up in what is going on with account ratcheting.  I don't think we can design this interface without clear agreements on what that looks like in the future as well.

**Comment by m-branson** [comment]
I'm replying to various people in this message, **please read in full**. I encourage everyone to continuing reading & feedback until we are aligned. (eg; I know this is a long issue, but it's worth the async effort!) 

_Additionally, in the future, I will use Discourse for this type of alignment. Honestly I had no clue it would go this deep._

@fiji-flo:
>However this makes everything a little bit messy. I my personal ideal world the IdP is part of Mozillians / our profile management tool.

Meaning you think it should be _integrated_ in the profile tool or it would be okay if it was _linked_ from the profile tool?

>This would mean every common voice user also has a profile (which is not visible to anyone unless the user upgrades it to a Mozillians profile).

This statement leads me to believe you mean IdP would be _integrated_; though I'm not sure what you mean by _upgrading to a Mozillians profile_?

@gdestuynder:
>Could the "IdP portal" simply be Mozillians.org with the UX refresh, so that it looks like its integrated?

It will be redesigned to look 'part of the IAM family' regardless. What I'm trying to get us to is: where should the IdP function live technically? Some options:
1. As an _integrated part_ of the Mozillians.org Profile; to edit you must view a Mozillians.org profile
2. As an _entity_ of Mozillians.org that is link-able from the Mozillians.org Profile page; to edit users can route via their Mozillians.org Profile page _or_ from the user menu _or_ from a Common Voice profile.
3. As an _entity not part of Mozillians.org_ that is link-able in the same way as No. 2

>We only allow one entity to edit a profile field (ie each field is owned and controlled by one entity), which means if common voice had its own way of setting the user's account verifications, then mozillians.org wouldn't be able to do it anymore.

This is because otherwise we run into issues where different parties disagree on the field/attribute values.

Understood. (using the Common Voice example again) Currently for Common Voice we are only using IAM for NLX to authenticate. Once Profile v2 is ready, I believe we are looking to revisit whether or not Common Voice profiles become integrated with Profile v2 and ultimately managed by Mozillians.org. 

This is still a point of contention: 
- some desire this integration which would direct users to Mozillians.org for all Profile v2 level information managment;
- some prefer to keep all profile management at the RP level & route to portal type environments to manage authentication (NLX) & verifed accounts (IdP entity)

^ @hmitsch & @flekkowich does this accurately capture both perspectives?

@andrewkrug 
>When I see the Mozillians.org IdP method diagram and it contains so many login methods that adds a lot of code that will have to be maintained over time.

From what I understand, these are the same login methods we are currently supporting. Is it the verified accounts specifically that concern you?

>Today we only have need to verify Github, Bugzilla, and FxA. There is no current proposed use case for a secure assertion including a Mozillian's twitter account for example.

I'm proposing one, why would we not verify accounts for instances like Slack & Twitter if we _are_ going to verify an account like Github & Bugzilla which just happen to _also_ be IdPs?
- Use case that comes to mind: the user who logs in with FxA, but wants to connect Github & Bugzilla as verified accounts where they contribute. We force a verification for those accounts but not for accounts like Slack & Twitter. Which, while also not login accts (IdPs) are important accounts to the user.
- How do we explain the difference? Why not create a rule now where verification is a requirement to connect accounts?

>A lot of this is also caught up in what is going on with account ratcheting. I don't think we can design this interface without clear agreements on what that looks like in the future as well.

Agreed. @hmitsch I believe you have a separate message coming this week (via Discourse?) that will further align us in the account ratcheting realm.

**Comment by m-branson** [comment]
This thread and a further mtg today with @hmitsch @flekkowich have helped pave a way forward where we look to explore the use of a `central profile utility` and a `central login utility`. 

The current UX architecture map is [here](https://realtimeboard.com/app/board/o9J_kzEB-Ls=/). I will be making some edits based on comments by @gdestuynder and will start to map the future ideal proposal this week. I welcome further comments on the current map for those that have time.
